ut/client_ut.cpp

```diff
@@ -631,6 +658,71 @@ TEST_P(ClientCase, Decimal) {
     });
 }
 
+// Test roundtrip of DateTime64 values
+TEST_P(ClientCase, DateTime64) {
+    const auto & server_info = client_->GetServerInfo();
+    if (versionNumber(server_info) < versionNumber(20, 1)) {
+        std::cout << "Test is skipped since server '" << server_info << "' does not support DateTime64" << std::endl;
+        return;
+    }
+
+    Block block;
+    client_->Execute("DROP TABLE IF EXISTS test_clickhouse_cpp.datetime64;");
+
+    client_->Execute("CREATE TABLE IF NOT EXISTS "
+            "test_clickhouse_cpp.datetime64 (dt DateTime64(6)) "
+            "ENGINE = Memory");
+
+    auto col_dt64 = std::make_shared<ColumnDateTime64>(6);
+    block.AppendColumn("dt", col_dt64);
+
+    // Empty INSERT and SELECT
+    client_->Insert("test_clickhouse_cpp.datetime64", block);
+    client_->Select("SELECT dt FROM test_clickhouse_cpp.datetime64",
+        [](const Block& block) {
+            ASSERT_EQ(0U, block.GetRowCount());
+        }
+    );
+
+    const std::vector<Int64> data{
+        -1'234'567'890'123'456'7ll, // approx year 1578
+        -1'234'567'890'123ll,       // 1969-12-17T17:03:52.890123
+        -1'234'567ll,               // 1969-12-31T23:59:58.234567
+        0,                          // epoch
+        1'234'567ll,                // 1970-01-01T00:00:01.234567
+        1'234'567'890'123ll,        // 1970-01-15T06:56:07.890123
+        1'234'567'890'123'456'7ll   // 2361-03-21T19:15:01.234567
+    };
+    for (const auto & d : data) {
+        col_dt64->Append(d);
+    }
+
+    block.RefreshRowCount();
+
+    // Non-empty INSERT and SELECT
+    client_->Insert("test_clickhouse_cpp.datetime64", block);
+
+    size_t total_rows = 0;
+    client_->Select("SELECT dt FROM test_clickhouse_cpp.datetime64",
+        [&total_rows, &data](const Block& block) {
+            total_rows += block.GetRowCount();
+            if (block.GetRowCount() == 0) {
+                return;
+            }
+
+            ASSERT_EQ(1U, block.GetColumnCount());
+            if (auto col = block[0]->As<ColumnDateTime64>()) {
+                ASSERT_EQ(data.size(), col->Size());
```

Comment:
Does this assume, that you always read an entire `data`-sized block here?

Reply:
good note, fixed